Support CUDA GPU accelerated template matching

din14970 opened the pull request:
---
name: GPU template matching and polar image transformation
about: First ventures into GPU accelerated analysis of 4D-STEM data
---

**Checklist**
- [ ] Updated CHANGELOG.md
- [x] (if finished) Requested a review (from pc494 if you are unsure who is most suitable)

**What does this PR do? Please describe and/or link to an open issue.**
Since I found a bug in the template matching approach, which should be resolved with #740, I looked again into the details. This PR continues from that branch, so #740 is more urgent. 

There is a lot of untapped potential in analyzing 4D-STEM data with the GPU as most problems are embarrassingly parallel and involve "dumb" linear algebra operations. As a test-case, since I was looking into template matching, I tried to re-implement it using cupy and numba.cuda as described in #739. It also served as a little test to see how well all these components work together with dask; from my experience it seems pretty seamless - dask arrays can also be cupy arrays it computes.

The template matching approach involves a conversion to polar coordinates of all the images and templates, then shifting all the templates across the theta axis of the polar image and calculating the correlation. Both the polar conversion and template shifting approach were parallelized on GPU with incredible performance gains. For the polar transform I could get a speed up of 40-400x depending on the image size (scikit-image's warp-polar against the implementation in this PR), for the indexing of a single pattern I could achieve a speed up of 5-10 depending on the size of the template library compared to 40 cpu cores. Larger data generally yields larger relative speed gains when using the GPU.

In order to support analysis of large 4D-STEM datasets, pyxem could get a real edge by thinking about GPU and HPC implementations of the algorithms a priori. Even in the upstream libraries like scikit-image this is still a point of active development. I would welcome a discussion on where else in the code-base we could make an impact. I'm pretty sure vector based analysis and peak finding may also seriously benefit from the GPU.

Another thing I changed is I removed the argument for parallelizing the polar transform during the indexation process. This was previously implemented with a hacky `njit` function that includes `objmode` blocks to do the actual polar transform. There seemed to be limited gain in using this function over the simple python for-loop implementation, so I default to this unless the GPU is used.

**Describe alternatives you've considered**
I was already parallelizing the heck out of the problem with dask and numba, which worked fine on a cluster or heavy duty workstation. With CUDA we can gain 1-2 orders of magnitude in speed up, which gives us a serious edge over commercial software, and maybe warrants a little publication - would anyone like to contribute on this point?

**Are there any known issues? Do you need help?**
I have absolutely no idea how to deal with unit testing. Does Github actions support GPU code? How can one create unit tests that only run when there is a CUDA enabled GPU on the system? Or is there a way to emulate a GPU but run it on CPU for unit testing?

**Work in progress?**
- [x] Needs quite a bit more testing, optimization and benchmarking. Just opening this up already for any feedback.

ericpre commented:
Have you tried most of the code which would dispatch to either numpy/scipy/scikit-image or cupy? Except for the cuda kernel, where you will need to branch the code to use either the CPU or GPU implementation, for the rest it should be possible to use the same code for both GPU and CPU.

Are you using https://github.com/hyperspy/hyperspy/pull/2670? If so, I would curious to know how well it works.

Regarding testing, at the moment (and most likely for a long time), the GPU part of the tests suite will need to run locally. You can have a look at https://github.com/hyperspy/hyperspy/pull/2670/files#diff-bad77c9b462b993417fd202c8b2fbac32dae859871f23a23ba99bb73e56dbb94 for an example.

din14970 commented:
@ericpre 
> Have you tried most of the code which would dispatch to either numpy/scipy/scikit-image or cupy? Except for the cuda kernel, where you will need to branch the code to use either the CPU or GPU implementation, for the rest it should be possible to use the same code for both GPU and CPU.

Does cupy default to numpy/CPU if there is no GPU, no right? `asarray` and `asnumpy` are still quite specific. I guess there are workarounds for making code work either on numpy or cupy but I haven't delved too deeply, I mainly wanted to get it working first. I should really spend some time looking at what you did. 

The biggest challenge is that the CPU code relies mostly on numba JIT compiled code, which as far as I'm aware does not support cupy arrays. Also scikit-image does not currently support cupy, so I re-implemented `warp_polar` from scratch for GPU, but use the scikit-image function for the CPU branch. There may be alternative ways to do everything of course.

> Are you using hyperspy/hyperspy#2670? If so, I would curious to know how well it works.

No, everything is custom using dask, numba and cupy. Haven't had a chance to play with this yet, I only have a GPU on the cluster and it's a bit more combersome to experiment than with a local GPU. Probably when things like `map` get improved upstream, the things in this PR can be significantly simplified.

> Regarding testing, at the moment (and most likely for a long time), the GPU part of the tests suite will need to run locally. You can have a look at https://github.com/hyperspy/hyperspy/pull/2670/files#diff-bad77c9b462b993417fd202c8b2fbac32dae859871f23a23ba99bb73e56dbb94 for an example.

Too bad, I was really hoping there would be some GPU virtualisation to test GPU code on CPU. Thanks for the link!

ericpre commented:
> > Have you tried most of the code which would dispatch to either numpy/scipy/scikit-image or cupy? Except for the cuda kernel, where you will need to branch the code to use either the CPU or GPU implementation, for the rest it should be possible to use the same code for both GPU and CPU.
> 
> Does cupy default to numpy/CPU if there is no GPU, no right? `asarray` and `asnumpy` are still quite specific. I guess there are workarounds for making code work either on numpy or cupy but I haven't delved too deeply, I mainly wanted to get it working first. I should really spend some time looking at what you did.

There is the array protocol which dispatch array to the relevant library and there are pattern to follow to use this protocol as I mentioned in https://github.com/pyxem/pyxem/issues/737#issuecomment-791011799 and https://github.com/hyperspy/hyperspy/pull/2670. The latter contains the relevant links. If you have a look at https://github.com/hyperspy/hyperspy/pull/2670, you will notice that the idea is to have the data to live as cupy array and use `asnumpy` only when necessary, typically plotting.

> The biggest challenge is that the CPU code relies mostly on numba JIT compiled code, which as far as I'm aware does not support cupy arrays. Also scikit-image does not currently support cupy, so I re-implemented `warp_polar` from scratch for GPU, but use the scikit-image function for the CPU branch. There may be alternative ways to do everything of course.

Similarly, there is the CUDA array interface, which is supported by both cupy and numba, see https://docs.cupy.dev/en/stable/reference/interoperability.html. It seems that your `wrap_polar` use numpy and scipy, so it should be possible to write if for both GPU and CPU using the relevant array protocols.

> > Are you using [hyperspy/hyperspy#2670](https://github.com/hyperspy/hyperspy/pull/2670)? If so, I would curious to know how well it works.
> 
> No, everything is custom using dask, numba and cupy. Haven't had a chance to play with this yet, I only have a GPU on the cluster and it's a bit more combersome to experiment than with a local GPU. Probably when things like `map` get improved upstream, the things in this PR can be significantly simplified.

Well, this is exactly, where it is interesting. As mentioned in https://github.com/hyperspy/hyperspy/pull/2670, if you are using dask>=2021.03.0 and numpy >=1.20, which both implement NEP35 (see https://github.com/dask/dask/pull/6738), the code should be very very similar for CPU and GPU, except for using cuda kernel, obviously, therefore my suggestion earlier in this PR - sorry if this was not clear.

din14970 commented:
@ericpre ok finally had a look at hyperspy/hyperspy#2670, indeed looks great! Especially your small utility functions look very useful for reducing the branching I've got going on here. When do you suspect this might make it into a hyperspy release; I see you are targeting 1.7? Would you mind if I temporarily copy some of those functions into pyxem utils? Since this is a rather specific problem, and it seems to be working properly, I think this PR may make it into pyxem sooner. Also your unit tests are good examples, thanks!

One thought: the `to_gpu` function copies the entire dataset to the GPU, even if you pass in a dask array. From my understanding, `cp.as_array` will compute the result of the dask array and attempt to store it into GPU memory. For most datasets and GPUs I suppose that won't be a problem but I'd like to be conservative and have the option to keep the dataset on disk and only copy to GPU on a chunk-wise basis via dask. To be fair, I'm not sure whether dask "knows" when the memory of the GPU is full during its graph execution. In any case, the workflow I envisioned and tried to implement for template matching is something like:

![image](https://user-images.githubusercontent.com/25320842/111646426-f0216f80-8801-11eb-8ba4-085927a6451c.png)

From what I've tested, dask seems to work fine with cupy, so `da.from_array(cp.array([1, 2, 3]))` will produce a cupy array on compute. But reducing duplication between the CPU/GPU implementations of `warp_polar` and `_index_chunk` should be possible following your example from hyperspy/hyperspy#2670.

din14970 commented:
Still somewhat problematic is that `scipy.ndimage` functions don't accept `cupy` arrays, while `cupyx.scipy.ndimage` functions don't accept `numpy` arrays. So I can't seem to get around doing rather silly things like:

```
 if is_cupy_array(image):
        dispatcher = cp
        map_function = ndigpu.map_coordinates
else:
    dispatcher = np
    map_function = ndimage.map_coordinates
```
Ideal would be if scipy and scikit-image also worked with cupy arrays.

ericpre commented:
> One thought: the `to_gpu` function copies the entire dataset to the GPU, even if you pass in a dask array. From my understanding, `cp.as_array` will compute the result of the dask array and attempt to store it into GPU memory.

Indeed, this method is a convenience function for in-memory computation. For lazy signal, it is most likely better to delegate this task to dask. Most of the time, hyperspy will not need to care if the data is a dask array. For the `deconvolve` function in this [dask blog](https://blog.dask.org/2020/11/12/deconvolution) can work on numpy, cupy, dask/numpy or dask/cupy array.
Would you like to put your comments on that PR, so that we don't forget about it?

> Still somewhat problematic is that `scipy.ndimage` functions don't accept `cupy` arrays, while `cupyx.scipy.ndimage` functions don't accept `numpy` arrays. So I can't seem to get around doing rather silly things like:
> 
> ```
>  if is_cupy_array(image):
>         dispatcher = cp
>         map_function = ndigpu.map_coordinates
> else:
>     dispatcher = np
>     map_function = ndimage.map_coordinates
> ```
> 
> Ideal would be if scipy and scikit-image also worked with cupy arrays.

Indeed, `cupy` provides functionality to return the suitable module (`cupyx.scipy` or `scipy`) but this doesn't seem to work with the `ndimage` submode.

din14970 commented:
@pc494 just as an update, in trying to fix all the CUDA related stuff I did a major overhaul of the entire code under the hood of the fast template matching to reduce duplication, make it hopefully more maintainable and generalized. Now most functions should just work on cupy arrays. I also simplified some of the CPU based functions, for example, I removed the option to simultaneously find the central beam in the orientation mapping as there were already too many kwargs to take care of. I left this option in the functions for indexing a single pattern. 

It seems that the overhaul has also fixed the issues with the tests on different platforms, no idea why or how. I also copied most of the tests to check whether it will work for cupy arrays (images on the GPU) and will see how well it works tomorrow on the cluster. On paper of course coverage drops as I have to test explicitly for cupy in many places, which is not available on the CI, and so some blocks can never run. I'll try to see if I can get close to 100% coverage when a GPU is available on the system. But once I have done those checks I would say this is more or less ready.

din14970 commented:
![Screenshot 2021-05-13 at 19 44 34](https://user-images.githubusercontent.com/25320842/118164902-ed1eb600-b423-11eb-974c-c3fff57bd89a.png)

All the tests now pass for cupy arrays also. With that coverage should be as complete as possible.

din14970 commented:
Sorry for dropping off the face of the earth for a while. I haven't changed anything on this so it should be good for review yes, and since it's pretty massive I suppose it will take you a while to get through -- sorry. As @hakonanes says, yes you will need a CUDA enabled graphics card to try it out; I did all testing on a GPU cluster.

din14970 commented:
Sorry for dropping off the face of the earth for a while. I haven't changed anything on this so it should be good for review yes, and since it's pretty massive I suppose it will take you a while to get through -- sorry. As @hakonanes says, yes you will need a CUDA enabled graphics card to try it out; I did all testing on a GPU cluster.

pc494 commented:
I'm going to set up a pull request to your branch @din14970 addressing some small things.

pc494 commented on pyxem/utils/cuda_utils.py:
```diff
@@ -0,0 +1,133 @@
+            ]
+            * sim_i[template, spot]
+        )
+    correlation[template, shift] = tmp
```
this looks very weird to me, could you explain?

din14970 replied:
The advantage of a GPU is doing as much as possible in parallel on separate threads, since there are thousands of individual cores. For some image (converted to polar coordinates) we want to know the best matching template and the best-matching in-plane angle of said template. To find the optimal solution, we have to calculate the correlation for each template at each in-plane angle; this can all be done independently and in parallel. So we define a 2D "grid" of GPU threads, each one responsible for one template and one in-plane angle. This function is a "kernel" that is run on each of these threads in parallel. A thread knows about it's own "position" in the grid, which is information we can use to put the result in the correct place of the output array. With the line
```python
template, shift = cuda.grid(2)
```
each thread will know what position it is in in the grid. I will then map the first coordinate onto the template index and the second coordinate onto the "shift" (= in plane rotation) coordinate in the output array `correlation`. 

 Typically we want a 1-to-1 correspondence between the grid of threads and the output array. However, due to the architecture of the GPU, it may be that the grid is slightly larger and that some threads should do nothing (certainly not update any values). To ensure a thread doesn't do anything if it's outside the boundaries of the 2D array we want to calculate there's the lines:
```python
if template >= correlation.shape[0] or shift >= correlation.shape[1]:
        return
```
Then we set up the actual calculation each thread must perform. A temporary variable is declared which will be our result. 
```
tmp = 0.0
```
The value is stored in a local bit of memory for each thread. Each thread must then loop over all the diffraction spots in the template it is responsible for.  
```python
for spot in range(sim_r.shape[1]):
```
We loop over the the maximum number of spots but break the loop once it encounters a spot where `r=0` (not all templates have an equal number of spots, so we have padded with 0's) . 
```python
if sim_r[template, spot] == 0:
            break
```
Then to actually calculate the contribution  of each spot to the correlation we have:
```python
tmp += (
            polar_image[
                (sim_t[template, spot] + shift) % polar_image.shape[0],
                sim_r[template, spot],
            ]
            * sim_i[template, spot]
        )
```
which basically says: get me the pixel in the polar image that corresponds to the coordinate of the right spot, shifted along the theta direction by `shift` (modulo 360), and multiply it by the template spot intensity. This is basically our dot product.

When we have looped over all the spots in the template, we can place our calculated value in the right place of the output array.

After running the kernel over all shifts and templates (which happens in parallel) we find the maximum correlation by taking the maximum over the shift axis (happens elsewhere). 

Does this explain it adequately?

One could also do this simultaneously and instead of a 2D grid, have each thread be responsible for one template - a 1D grid, and calculating the optimum shift during the kernel's execution. However unintuitive, my benchmarks showed that this is actually slower that calculating the entire array of possibilities and then calculating the maximum element in each row (basically a bit map-reduce like).

I've thought of additional ideas to further optimize the entire procedure but for now I think this works well enough.

hakonanes replied:
Especially the explanation of the use of `cupy.grid()` was clarifying for me, thanks!

pc494 replied:
That helps a lot, it's just very strange to see a "function" that doesn't end with a `return` - but as far as I understand it that's very standard for this sort of CUDA coding?

din14970 replied:
Yep, cuda kernels can not allocate global memory as far as I understand, so you have to pass it a reference to the output array which should already be pre-allocated somehow. I made a little tutorial on cuda programming in python in case you're interested: https://youtu.be/9bBsvpg-Xlk

pc494 replied:
Great, thanks for this

pc494 commented on pyxem/utils/expt_utils.py:
```diff
@@ -575,11 +586,17 @@ def find_beam_center_blur(z, sigma):
     -------
     center : np.array
-        np.array [y, x] containing indices of estimated direct beam positon.
+        np.array [x, y] containing indices of estimated direct beam positon.
```
is this a documentation change, or a behaviour change? Does it have to be like this?

din14970 replied:
I didn't functionally change the code, only made it capable of handling cupy arrays. I think this was simply a mistake in the documentation -> the `[::-1]` reverses the indices, so the output will be `[x, y]` not `[y,x]`

pc494 replied:
simple, will mark as resolved

pc494 reviewed (commented):
PR to your branch coming shortly, just with formatting stuff. Couple of questions and then we can get this merged I hope!

pc494 commented:
The above mentioned PR is now in as: https://github.com/din14970/pyxem/pull/1

pc494 commented:
Failure is unrelated, merge.
